### PR TITLE
Model discrete sampling geometry datasets

### LINKED ISSUE #60: Model discrete geometry datasets
Streamflow models create non-gridded, station-centric data, called "[discrete sampling geometry](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch09.html)" in CF terminology. 

`modelmeta` needs to model discrete sampling geometries as well as gridded sampling geometries.

### PR DESCRIPTION
**Pull request by rod-glover**
Resolves #60 

1. Includes Alembic schema+data migration. Both upgrade and downgrade have been tested on a SQLite db but not a Postgres db.

1. Modifies indexer and tests to work with new ORM/schema, but only for gridded datasets, which were the kind handled originally. Indexing DSG TS datasets will be the [subject](https://github.com/pacificclimate/modelmeta/issues/63) of another PR.

1. Cleans up and PEP8-ifies v2 ORM code, which was a little untidy.

1. Adds tests specifically for new, polymorphic ORM for discrete sampling and gridded sampling geometries. These may seem a little superfluous, but they are a way of ensuring that we correctly understand how SQLAlchemy handles polymorphism/inheritance. Turns out it's pretty straightforward.

TODO:

- [x] In database migration, either 
  - [ ] remove all `DataFile.*_dim_name` attributes, or 
  - [x] make them all nullable (some aren't) so that they can be nulled out for DSG TS files, to which these attrs are irrelevant.

- [x] ? Test migration against a Postgres database? Can we (Matthew, heh) make a copy of `ce_meta` easily? That would be best.

### CODE REVIEW

File: alembic/versions/12f290b63791_handle_variant_sampling_geometries.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        sa.MetaData(bind=op.get_bind()),
+        autoload=True
+    )
+    dfvs.delete().where(dfvs.c.geometry_type == 'dsg_time_series')
+
+
+
+def downgrade():
+    # This downgrade will dump all discrete geometry data.
+    # TODO: Add user confirmation of downgrade?
```
Comment (rod-glover) [inline alembic/versions/12f290b63791_handle_variant_sampling_geometries.py]: @jameshiebert what do you think?

File: tests/modelmeta/test_orm.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    sesh.add(dfv_dsg_time_series_1)
+    sesh.flush()
+    print()
+    print(dfv_dsg_time_series_1)
+    assert dfv_dsg_time_series_1.geometry_type == 'dsg_time_series'
+    assert data_file_1.data_file_variables == [dfv_dsg_time_series_1]
+    assert variable_alias_1.data_file_variables == [dfv_dsg_time_series_1]
+
+
+def test_station(test_session_with_empty_db, station_1):
```
Comment (rod-glover) [inline tests/modelmeta/test_orm.py]: Remove this test. It's too simple.

File: modelmeta/v2.py [hunk 3]
```diff
@@ -23,17 +54,21 @@
...
+        Integer, 
+        ForeignKey('time_sets.time_set_id'), 
+        primary_key=True, 
+        nullable=False
+    )
 
     def __repr__(self):
-        return '{}({}, {}, {}, {})'.format(self.__class__.__name__,
-                self.time_idx, self.time_end, self.time_start, self.time_set_id)
+        return obj_repr('time_set_id time_idx time_start time_end', self)
```
Comment (jameshiebert) [inline modelmeta/v2.py]: Yeah, that's handy.

File: modelmeta/v2.py [hunk 4]
```diff
@@ -71,78 +111,249 @@ def __str__(self):
...
+        'polymorphic_on': geometry_type
+    }
+
+    def __repr__(self):
+        return obj_repr(
+            'id geometry_type derivation_method variable_cell_methods '
+            'netcdf_variable_name disabled range_min range_max', self)
+
+
+class DataFileVariableDSGTimeSeries(DataFileVariable):
```
Comment (jameshiebert) [inline modelmeta/v2.py]: Could probably use a docstring here to elaborate on what DSG stands for.

**Comment by rod-glover** [comment]
1. Yes, sorry about mixing in the PEP-8 fixup. I can remove it if you like and put it in a separate PR. Shouldn't be very difficult.

2. So far, migrations don't have unit tests. The reason is that within a PR, the ORM matches the final migrated state of the database, not an earlier state. The unit tests (relying on `testing.postgresql`) just create the database from the (current) ORM. There are a couple of ways we could create a unit test environment for the upgrade migrations:
   1. Stash an old version of the ORM, create a test database based on that, populate it with some test data, migrate, run unit tests.
   1. Populate and dump a Postgres database in the old state, load it up in, migrate, run unit tests.
   1. Both of those solutions have always seemed fairly effortful to me, and so I rely on testing against local databases, whether SQLite or Postgres. I am currently making a copy of ce_meta locally so that I have something safe to test against.

3. For the downgrade migrations, it's easier, since you're downgrading from a latest version of the ORM.

**Comment by rod-glover** [comment]
OK, embarassing, there's [this](http://alembic-verify.readthedocs.io/en/latest/). I'll look into it and any similar  packages.

**Comment by jameshiebert** [comment]
> OK, embarassing, there's this. I'll look into it and any similar packages.

Wouldn't call that embarrassing. I didn't know about that, and it looks like a good resource. To boot, I hadn't thought through the fact that to test the database, your initial instance will be the current ORM schema. In any case, checking out ``alembic-verify`` sounds like a good path.

>  I can remove it if you like and put [the PEP8 migration] in a separate PR. 

No don't worry about it. Just keep it in mind for next time.

**Comment by jameshiebert** [comment]
LGTM. Are you happy with this at present @rod-glover ?

**Comment by rod-glover** [comment]
@jameshiebert : One small addition: Migration now makes` data_files.x_dim_name`, .`y_dim_name` nullable so that we don't have to fill in nonsense values for them for DSG data files. We didn't really make a decision whether or not to remove these (I'd like to), so this seemed the prudent choice at the moment. We can migrate those columns out later if we decide to. I'll put in an issue about that.

I am happy with this now and ready to merge if you are.

**Pull request merged**